src/Simplex/Messaging/Agent/Protocol.hs

```diff
@@ -135,44 +154,150 @@ instance TestEquality SAParty where
   testEquality SClient SClient = Just Refl
   testEquality _ _ = Nothing
 
-data ACmd = forall p. ACmd (SAParty p) (ACommand p)
+-- | SMP agent protocol entity types
+data EntityTag = Conn_ | OpenConn_ | Broadcast_ | AGroup_
+
+data Entity :: EntityTag -> Type where
+  Conn :: ByteString -> Entity Conn_
+  OpenConn :: ByteString -> Entity OpenConn_
+  BroadCast :: ByteString -> Entity Broadcast_
+  AGroup :: ByteString -> Entity AGroup_
+
+-- deriving instance Eq (Entity t)
+
+deriving instance Show (Entity t)
+
+instance IsString (Entity Conn_) where
+  fromString = Conn . B.pack
+
+data AnEntity = forall t. AE (Entity t)
+
+data ACmd = forall (p :: AParty) (c :: ACmdTag). ACmd (SAParty p) (ACommand p c)
 
 deriving instance Show ACmd
 
+data APartyCmd (p :: AParty) = forall c. APartyCmd (ACommand p c)
+
+instance Eq (APartyCmd p) where
+  APartyCmd c1 == APartyCmd c2 = isJust $ testEquality c1 c2
+
+deriving instance Show (APartyCmd p)
+
+type family EntityCommand (t :: EntityTag) (c :: ACmdTag) :: Constraint where
+  EntityCommand Conn_ NEW_ = ()
+  EntityCommand Conn_ INV_ = ()
+  EntityCommand Conn_ JOIN_ = ()
+  EntityCommand Conn_ CON_ = ()
+  EntityCommand Conn_ SUB_ = ()
+  EntityCommand Conn_ SUBALL_ = ()
+  EntityCommand Conn_ END_ = ()
+  EntityCommand Conn_ SEND_ = ()
+  EntityCommand Conn_ SENT_ = ()
+  EntityCommand Conn_ MSG_ = ()
+  EntityCommand Conn_ OFF_ = ()
+  EntityCommand Conn_ DEL_ = ()
+  EntityCommand Conn_ OK_ = ()
+  EntityCommand Conn_ ERR_ = ()
+  EntityCommand OpenConn_ NEW_ = ()
+  EntityCommand Broadcast_ NEW_ = ()
+  EntityCommand AGroup_ NEW_ = ()
+  EntityCommand _ NEW_ = ()
```

Comment:
why do we need a NEW command of any entity if we already have it separately for each entity?